backend/routes/graph.py

```diff
@@ -29,166 +35,308 @@ def subgraph():
+#     for path in paths:
+#         root.insert(path)
+#
+#     # Use StringIO to create a CSV string
+#     output = io.StringIO()
+#     writer = csv.writer(output)
+#     writer.writerow(['level', 'node'])  # Writing header
+#     for line in generate_csv_lines(root):
+#         writer.writerow(line)
+#
+#     # Get the CSV string from StringIO
+#     csv_string = output.getvalue()
+#     output.close()
+#
+#     # csv_string contains the CSV data
+#     return csv_string
+
+
+# def connected_subgraph_from_nodes(nodes, G, G_undirected=None, verbose=False):
+#     # what this fails to do is fill in gaps if the subgraph is already connected
+#     timer = get_timer('  connected_subgraph_from_nodes')
+#     if not G_undirected:
+#         verbose and timer(f'convert G to undirected')
+#         G_undirected = G.to_undirected()
+#
+#     verbose and timer(f'subgraph for {len(nodes)} nodes')
+#     sg = G.subgraph(nodes)
+#     verbose and timer('get roots')
+#     roots = get_roots(sg)
+#     pairs = [(roots[i], roots[i+1]) for i in range(0, len(roots) - 1)]
+#     connected_nodes = set(nodes)
+#     verbose and timer(f'get shortest paths for {len(pairs)} pairs of root nodes')
+#     for p in pairs:
+#         try:
+#             verbose and timer(f'   path between {p[0]} and {p[1]}')
+#             path = nx.shortest_path(G_undirected, *p)
+#             connected_nodes.update(path)
+#         except Exception as err:
+#             warnings.warn(str(err))
+#     verbose and timer(f'subgraph for {len(connected_nodes)} connected nodes')
+#     sgc = G.subgraph(connected_nodes)
+#     verbose and timer('done')
+#     return sgc
+#
+#
+# def sort_by_most_descendants(G, nodes):
+#     key = lambda n: len(nx.descendants(G, n))
+#     return sorted(nodes, key=key, reverse=True)
+#
+# def graph_to_json(G, sorted=False):
+#     root_nodes = [node for node in G.nodes if G.in_degree(node) == 0]
+#     if sorted:
+#         root_nodes = sort_by_most_descendants(G, root_nodes)
+#     hierarchies = {root: create_hierarchy(G, root, sorted) for root in root_nodes}
+#     return hierarchies
+#
+#
+# def create_hierarchy(G, node, sorted=False):
+#     successors = list(G.successors(node))
+#     if len(successors) == 0:
+#         return None
+#     if sorted:
+#         successors = sort_by_most_descendants(successors)
+#     else:
+#         return {successor: create_hierarchy(G, successor, sorted) for successor in successors}
+#
+# def for_testing():
+#     G = nx.DiGraph()
+#     G.add_edges_from([(1, 2), (1, 3), (2, 4), (2, 5), (2,3), (2, 6), (3, 6), (9,2)])
+#     return G
+#
+#
+# def disconnected_subgraphs():
+#     G = nx.DiGraph([
+#         (1, 2), (2, 3), (3, 4),
+#         (1, 5), (5, 6), (6, 7),
+#         (10, 11), (11, 12), (12, 6)
+#     ])
+#     components = G.subgraph([2,3,4,5,6,7])
+#     return G, components
+
+
+
+def generate_graph_edges():
+    with get_db_connection() as con:
+        # moving the sql to ddl-20-concept_graph.jinja.sql
```

Comment:
## `concept_graph` table
I don't see `ddl-20-concept_graph.jinja.sql` yet, I'm guessing that's still pending?

**Questions**
1. I'm curious to know what this table looks like and how it will be used?
2. Is it not redundant with `networkx`? Does it serve a different role?
3. I'm not seeing where this table gets populated. Has that code been written yet? Is this going to be populated as we were talking about, using the `concept_releationship` `is_a` edges?

Reply:
Forgot to add it earlier.  It gets [populated by the ddl](https://github.com/jhu-bids/TermHub/blob/fixing-concept-hierarchy/backend/db/ddl-20-concept_graph.jinja.sql) when [this code](https://github.com/jhu-bids/TermHub/blob/5aced3941d1dc4a39e0824568938d7ed4312aaa1/backend/routes/graph.py#L278C1-L318C20) runs, which happens when you start up the backend and the networkx file is missing.

Yes, it is redundant, but runs faster this way. And, btw, starting the backend is actually faster now.